### ISSUE TITLE
Moq1202/Moq1204 false positives: canonical `Raise` on non-generic `EventHandler` events is flagged, and unresolvable delegate types are treated as zero-parameter delegates

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1248]
## Summary

An internal code audit found two related defects in the shared event-signature validation used by Moq1202 (`Mock.Raise`) and Moq1204 (`Setup(...).Raises`):

1. **High — FP on canonical, documented Moq usage.** For a non-generic `event System.EventHandler Closed;`, the canonical call `mock.Raise(m => m.Closed += null, EventArgs.Empty);` is flagged Moq1202 "arguments should match" (too few arguments). Moq's `Raise(Action<T>, EventArgs)` overload auto-supplies the mock object as the `sender`, so this call is correct at runtime. The same false positive applies to any custom `(object sender, TArgs e)`-shaped delegate and to `.Raises(...)`/Moq1204.
2. **Medium — FP on mid-edit code.** When the event's delegate type cannot be analyzed (unresolved/error type mid-edit, non-delegate type), `EventSyntaxExtensions.GetEventParameterTypes` returns an empty array, which the consumer cannot distinguish from a genuine zero-parameter delegate. Every supplied argument is then flagged on code that is correct once the type resolves.

Both defects live in `src/Common/EventSyntaxExtensions.cs` and manifest through `src/Analyzers/RaiseEventArgumentsShouldMatchEventSignatureAnalyzer.cs` and `src/Analyzers/RaisesEventArgumentsShouldMatchEventSignatureAnalyzer.cs`.

### Moq semantics (verified against the Moq 4.8.2 and 4.18.4 packages used by the test suite)

Both Moq versions expose exactly two `Mock<T>.Raise` overloads (and matching `IRaise<T>.Raises` overloads):

- `Raise(Action<T> eventExpression, EventArgs args)` — invokes each attached handler as `handler(mock.Object, args)`. The **sender is supplied by Moq**; the caller passes only the `EventArgs`-derived payload.
- `Raise(Action<T> eventExpression, params object[] args)` — invokes each handler with `args` verbatim; the arguments must match the delegate's full `Invoke` parameter list positionally. Moq's own XML doc describes this overload as being "for a non-`System.EventHandler`-typed event", but it binds and works for `EventHandler`-typed events too when two arguments are passed.

Consequently, for any delegate shaped `(object sender, TArgs e)` where `TArgs` is `System.EventArgs` or derives from it (non-generic `EventHandler`, `EventHandler<TArgs>` with an `EventArgs`-derived `TArgs`, custom delegates), **both** of these are legal:

- 1-argument form: `mock.Raise(m => m.E += null, args)` where `args` is `EventArgs`-compatible (binds the `EventArgs` overload; Moq supplies the sender), and
- 2-argument form: `mock.Raise(m => m.E += null, sender, args)` (binds the `params object[]` overload).

For every other delegate, only the exact positional form is legal.

### Defect 1 — non-generic `EventHandler` falls through to the raw Invoke signature

`src/Common/EventSyntaxExtensions.cs:61-75`:

```csharp
internal static ITypeSymbol[] GetEventParameterTypes(ITypeSymbol eventType, KnownSymbols knownSymbols)
{
    if (eventType is not INamedTypeSymbol namedType)
    {
        return [];
    }

    // Try different delegate type handlers in order of specificity
    ITypeSymbol[]? parameterTypes =
        TryGetActionDelegateParameters(namedType, knownSymbols) ??
        TryGetEventHandlerDelegateParameters(namedType, knownSymbols) ??
        TryGetCustomDelegateParameters(namedType);

    return parameterTypes ?? [];
}
```

`TryGetEventHandlerDelegateParameters` (`src/Common/EventSyntaxExtensions.cs:210-218`) only matches `EventHandler<T>` (it requires `namedType.TypeArguments.Length > 0` and compares against `KnownSymbols.EventHandler1`, i.e. ``System.EventHandler`1``). Non-generic `System.EventHandler` therefore falls through to `TryGetCustomDelegateParameters` (`src/Common/EventSyntaxExtensions.cs:225-241`), which reads the `Invoke` signature and returns `[object, EventArgs]`.

`ValidateEventArgumentTypes` (`src/Common/EventSyntaxExtensions.cs:25-33`) then requires an exact argument count:

```csharp
if (eventArguments.Length != expectedParameterTypes.Length)
{
    Location location = eventArguments.Length < expectedParameterTypes.Length
        ? invocation.GetLocation()
        : eventArguments[expectedParameterTypes.Length].GetLocation();

    context.ReportDiagnostic(CreateEventDiagnostic(location, rule, eventName));
    return;
}
```

**Concrete failure:** given `interface IFoo { event EventHandler Closed; }`, the canonical call

```csharp
mock.Raise(m => m.Closed += null, EventArgs.Empty);
```

supplies 1 argument against 2 expected parameters and is flagged Moq1202 on the whole invocation, even though it is the exact pattern from Moq's own documentation and runs correctly. The same happens for Moq1204 via `Setup(...).Raises(x => x.Closed += null, EventArgs.Empty)`, and for any custom `(object sender, XEventArgs e)` delegate event.

Note the existing special case for `EventHandler<T>` returns `[T]` — meaning the 1-argument form is accepted for `EventHandler<T>` but the equally legal 2-argument form (`mock.Raise(m => m.E += null, sender, args)`, which binds `params object[]`) is flagged as "too many arguments". That is a second, narrower FP with the same root cause.

### Defect 2 — unanalyzable types conflated with zero-parameter delegates

For an event whose delegate type does not resolve (verified against Roslyn 4.8: the event symbol still binds; its `Type` is an `INamedTypeSymbol` with `TypeKind.Error` and `DelegateInvokeMethod == null`):

- `TryGetActionDelegateParameters` → null, `TryGetEventHandlerDelegateParameters` → null, `TryGetCustomDelegateParameters` → null (`invokeMethod is null`), so `GetEventParameterTypes` returns `?? []` — indistinguishable from a real `event Action E;`.
- `TryGetEventMethodArguments` (`src/Common/EventSyntaxExtensions.cs:112`) does `expectedParameterTypes = GetEventParameterTypes(eventType, knownSymbols);` and returns `true`.
- `ValidateEventArgumentTypes` then flags **every** supplied argument (count mismatch: N supplied vs. 0 expected).

**Concrete failure:** mid-edit code where `UnresolvedDelegateType` is not yet defined:

```csharp
interface IFoo { event UnresolvedDelegateType Broken; }   // CS0246 on the type
...
mock.Raise(x => x.Broken += null, 42);                    // spurious Moq1202 on `42`
```

The same applies to `ITypeParameterSymbol` event types (only reachable in already-broken code, since `event TDel E;` is CS0066) via the `eventType is not INamedTypeSymbol` early return, and to non-delegate named types.

## Desired behavior

1. For any delegate whose `Invoke` signature is exactly `(object sender, TArgs e)` with `TArgs` being `System.EventArgs` or derived from it ("EventHandler-shaped": non-generic `EventHandler`, `EventHandler<TArgs>` with `EventArgs`-derived `TArgs`, and custom delegates), accept **both** the args-only form (validate the single argument against `TArgs`) and the full 2-argument form (validate positionally). This matches exactly the set of calls Moq accepts at runtime.
2. `EventHandler<TArgs>` where `TArgs` does **not** derive from `EventArgs` (e.g. `EventHandler<string>`) is not EventHandler-shaped: neither `Raise` overload can invoke it with one argument, so the full 2-argument form is required. (This is a small behavior improvement: today the analyzer accepts the 1-argument form for `EventHandler<string>`, which throws at runtime.)
3. When the event type is unanalyzable (`TypeKind.Error`, not a named type, not a delegate), extraction must **fail** (return `false` through the `Try*` chain) so the analyzers stay silent instead of pretending the delegate has zero parameters.
4. Detection stays symbol-based (ADR-001): the non-generic `System.EventHandler` and `System.EventArgs` types are resolved through `KnownSymbols`/`TypeProvider.GetOrCreateTypeByMetadataName`, and the shape check uses `SpecialType.System_Object` and symbol comparison against `System.EventArgs` — no string comparisons.

## Implementation plan

All paths repo-relative. `src/Common` has global usings for `System.Collections.Immutable`, `Microsoft.CodeAnalysis`, `Microsoft.CodeAnalysis.CSharp.Syntax`, `Microsoft.CodeAnalysis.Diagnostics`, `Moq.Analyzers.Common`, and `Moq.Analyzers.Common.WellKnown` (see `src/Common/GlobalUsings.cs`), so no new `using` directives are needed in the files below.

### 1. Add `EventHandler` and `EventArgs` to `src/Common/WellKnown/KnownSymbols.cs`

Mirror the existing registration style (cf. `EventHandler1` at line 48 and `Action0` at line 58). Insert after the existing `EventHandler1` property:

```csharp
    /// <summary>
    /// Gets the delegate <see cref="System.EventHandler"/> (the non-generic form).
    /// </summary>
    internal INamedTypeSymbol? EventHandler => TypeProvider.GetOrCreateTypeByMetadataName("System.EventHandler");

    /// <summary>
    /// Gets the class <see cref="System.EventArgs"/>.
    /// </summary>
    internal INamedTypeSymbol? EventArgs => TypeProvider.GetOrCreateTypeByMetadataName("System.EventArgs");
```

### 2. Replace `GetEventParameterTypes` with a `Try` method in `src/Common/EventSyntaxExtensions.cs`

Delete the three private helpers `TryGetActionDelegateParameters` (lines 199-202), `TryGetEventHandlerDelegateParameters` (lines 210-218), and `TryGetCustomDelegateParameters` (lines 225-241), and delete `GetEventParameterTypes` (lines 61-75). Replace with:

```csharp
    /// <summary>
    /// Attempts to get the parameter types for a given event delegate type.
    /// </summary>
    /// <param name="eventType">The event delegate type to analyze.</param>
    /// <param name="knownSymbols">Known symbols for type checking.</param>
    /// <param name="expectedParameterTypes">
    /// The parameter types expected by the event delegate:
    /// - For <see cref="System.Action"/>/<see cref="System.Action{T}"/> delegates: all generic type arguments
    /// - For all other delegates (including <see cref="System.EventHandler"/>, <see cref="System.EventHandler{T}"/>,
    ///   and custom delegates): the full <c>Invoke</c> parameter list.
    /// </param>
    /// <param name="senderCanBeOmitted">
    /// <see langword="true"/> when the delegate is "EventHandler-shaped" — its <c>Invoke</c> signature is exactly
    /// <c>(object sender, TArgs e)</c> with <c>TArgs</c> being <see cref="System.EventArgs"/> or derived from it.
    /// Moq's <c>Raise(..., EventArgs)</c>/<c>Raises(..., EventArgs)</c> overloads supply the mock object as the
    /// sender for such delegates, so the sender argument may be omitted at the call site.
    /// </param>
    /// <returns>
    /// <see langword="true"/> if the delegate signature could be analyzed; <see langword="false"/> for
    /// unresolved (error) types, non-named types (e.g. type parameters), and non-delegate types.
    /// </returns>
    internal static bool TryGetEventParameterTypes(
        ITypeSymbol eventType,
        KnownSymbols knownSymbols,
        out ITypeSymbol[] expectedParameterTypes,
        out bool senderCanBeOmitted)
    {
        expectedParameterTypes = [];
        senderCanBeOmitted = false;

        // Error types (mid-edit code) and non-named types (e.g. type parameters, arrays) cannot be
        // analyzed. Report failure instead of pretending the delegate has zero parameters.
        if (eventType.TypeKind == TypeKind.Error || eventType is not INamedTypeSymbol namedType)
        {
            return false;
        }

        // System.Action / System.Action<T>: the type arguments are the parameters.
        // NOTE: Action`2 and higher arities are intentionally resolved through the Invoke
        // signature below, exactly as before (IsActionDelegate only matches Action and Action`1).
        if (namedType.IsActionDelegate(knownSymbols))
        {
            expectedParameterTypes = namedType.TypeArguments.ToArray();
            return true;
        }

        // Any other delegate: read the Invoke signature.
        IMethodSymbol? invokeMethod = namedType.DelegateInvokeMethod;
        if (invokeMethod is null)
        {
            // Not a delegate type: unanalyzable.
            return false;
        }

        ImmutableArray<IParameterSymbol> parameters = invokeMethod.Parameters;
        ITypeSymbol[] types = new ITypeSymbol[parameters.Length];
        for (int i = 0; i < parameters.Length; i++)
        {
            types[i] = parameters[i].Type;
        }

        expectedParameterTypes = types;
        senderCanBeOmitted = IsEventHandlerShaped(namedType, parameters, knownSymbols);
        return true;
    }

    /// <summary>
    /// Determines whether a delegate is "EventHandler-shaped": <c>(object sender, TArgs e)</c> with
    /// <c>TArgs</c> being <see cref="System.EventArgs"/> or derived from it. Moq's
    /// <c>Raise(Action{T}, EventArgs)</c> and <c>Raises(Action{T}, EventArgs)</c> overloads invoke such
    /// delegates as <c>handler(mock.Object, args)</c>, so the sender argument may be omitted by callers.
    /// </summary>
    /// <param name="namedType">The delegate type.</param>
    /// <param name="invokeParameters">The delegate's <c>Invoke</c> parameters.</param>
    /// <param name="knownSymbols">Known symbols for type checking.</param>
    /// <returns><see langword="true"/> if the delegate is EventHandler-shaped; otherwise <see langword="false"/>.</returns>
    private static bool IsEventHandlerShaped(
        INamedTypeSymbol namedType,
        ImmutableArray<IParameterSymbol> invokeParameters,
        KnownSymbols knownSymbols)
    {
        // Fast path: the non-generic System.EventHandler delegate itself.
        if (knownSymbols.EventHandler is not null
            && SymbolEqualityComparer.Default.Equals(namedType, knownSymbols.EventHandler))
        {
            return true;
        }

        if (invokeParameters.Length != 2
            || invokeParameters[0].Type.SpecialType != SpecialType.System_Object)
        {
            return false;
        }

        INamedTypeSymbol? eventArgsType = knownSymbols.EventArgs;
        if (eventArgsType is null)
        {
            return false;
        }

        foreach (ITypeSymbol baseType in invokeParameters[1].Type.GetBaseTypesAndThis())
        {
            if (SymbolEqualityComparer.Default.Equals(baseType, eventArgsType))
            {
                return true;
            }
        }

        return false;
    }
```

Notes:
- `GetBaseTypesAndThis` already exists at `src/Common/ITypeSymbolExtensions.cs:13`.
- `NamedTypeSymbolExtensions.IsEventHandlerDelegate` (`src/Common/NamedTypeSymbolExtensions.cs:20`) loses its only production caller. Keep it — it is covered by `tests/Moq.Analyzers.Test/Common/NamedTypeSymbolExtensionsTests.cs`; do not delete it in this change.
- `IsActionDelegate` continues to be used and is unchanged.

### 3. Update `ValidateEventArgumentTypes` in `src/Common/EventSyntaxExtensions.cs` (currently lines 17-45)

Complete replacement:

```csharp
    /// <summary>
    /// Validates that event arguments match the expected parameter types.
    /// </summary>
    /// <param name="context">The analysis context.</param>
    /// <param name="eventArguments">The event arguments to validate.</param>
    /// <param name="expectedParameterTypes">The expected parameter types.</param>
    /// <param name="senderCanBeOmitted">
    /// When <see langword="true"/>, the delegate is EventHandler-shaped and Moq supplies the sender,
    /// so supplying one argument fewer than the parameter count (omitting the sender) is also valid.
    /// </param>
    /// <param name="invocation">The invocation expression for error reporting.</param>
    /// <param name="rule">The diagnostic rule to report.</param>
    /// <param name="eventName">The event name to include in diagnostic messages.</param>
    internal static void ValidateEventArgumentTypes(
        this SyntaxNodeAnalysisContext context,
        ArgumentSyntax[] eventArguments,
        ITypeSymbol[] expectedParameterTypes,
        bool senderCanBeOmitted,
        InvocationExpressionSyntax invocation,
        DiagnosticDescriptor rule,
        string? eventName = null)
    {
        int offset = 0;

        if (senderCanBeOmitted && eventArguments.Length == expectedParameterTypes.Length - 1)
        {
            // Moq's Raise/Raises(..., EventArgs) overloads supply the mock object as the sender for
            // EventHandler-shaped delegates. Validate the arguments against the post-sender parameters.
            offset = 1;
        }
        else if (eventArguments.Length != expectedParameterTypes.Length)
        {
            Location location = eventArguments.Length < expectedParameterTypes.Length
                ? invocation.GetLocation()
                : eventArguments[expectedParameterTypes.Length].GetLocation();

            context.ReportDiagnostic(CreateEventDiagnostic(location, rule, eventName));
            return;
        }

        for (int i = 0; i < eventArguments.Length; i++)
        {
            TypeInfo argumentTypeInfo = context.SemanticModel.GetTypeInfo(eventArguments[i].Expression, context.CancellationToken);
            ITypeSymbol? argumentType = argumentTypeInfo.Type;

            if (argumentType != null && !context.SemanticModel.HasConversion(argumentType, expectedParameterTypes[i + offset]))
            {
                context.ReportDiagnostic(CreateEventDiagnostic(eventArguments[i].GetLocation(), rule, eventName));
            }
        }
    }
```

This keeps existing spans intact: for `EventHandler<CustomArgs>` with a single wrong-typed argument, the argument itself is still flagged (the offset path validates it against `TArgs`), matching the existing markup in `tests/Moq.Analyzers.Test/RaiseEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs:54` and `tests/Moq.Analyzers.Test/Common/EventSyntaxExtensionsTests.cs:64`.

### 4. Update `TryGetEventMethodArguments` (currently lines 87-124) and `TryGetEventMethodArgumentsFromLambdaSelector` (currently lines 136-154)

Complete replacements:

```csharp
    /// <summary>
    /// Extracts arguments from an event method invocation.
    /// </summary>
    /// <param name="invocation">The method invocation.</param>
    /// <param name="semanticModel">The semantic model.</param>
    /// <param name="eventArguments">The extracted event arguments.</param>
    /// <param name="expectedParameterTypes">The expected parameter types.</param>
    /// <param name="senderCanBeOmitted">Whether the sender argument may be omitted (EventHandler-shaped delegates).</param>
    /// <param name="eventTypeExtractor">Function to extract event type from the event selector.</param>
    /// <param name="knownSymbols">Known symbols for type checking.</param>
    /// <returns><see langword="true" /> if arguments were successfully extracted; otherwise, <see langword="false" />.</returns>
    internal static bool TryGetEventMethodArguments(
        InvocationExpressionSyntax invocation,
        SemanticModel semanticModel,
        out ArgumentSyntax[] eventArguments,
        out ITypeSymbol[] expectedParameterTypes,
        out bool senderCanBeOmitted,
        Func<SemanticModel, ExpressionSyntax, (bool Success, ITypeSymbol? EventType)> eventTypeExtractor,
        KnownSymbols knownSymbols)
    {
        eventArguments = [];
        expectedParameterTypes = [];
        senderCanBeOmitted = false;

        SeparatedSyntaxList<ArgumentSyntax> arguments = invocation.ArgumentList.Arguments;

        if (arguments.Count < 1)
        {
            return false;
        }

        ExpressionSyntax eventSelector = arguments[0].Expression;
        (bool success, ITypeSymbol? eventType) = eventTypeExtractor(semanticModel, eventSelector);
        if (!success || eventType == null)
        {
            return false;
        }

        if (!TryGetEventParameterTypes(eventType, knownSymbols, out expectedParameterTypes, out senderCanBeOmitted))
        {
            // The delegate type could not be analyzed (e.g. an error type in mid-edit code).
            // Do not report diagnostics based on a guessed signature.
            return false;
        }

        if (arguments.Count > 1)
        {
            eventArguments = new ArgumentSyntax[arguments.Count - 1];
            for (int i = 1; i < arguments.Count; i++)
            {
                eventArguments[i - 1] = arguments[i];
            }
        }

        return true;
    }

    /// <summary>
    /// Extracts event arguments from an event method invocation using the standard
    /// lambda-based event type extraction pattern shared by Raise and Raises analyzers.
    /// </summary>
    /// <param name="invocation">The method invocation.</param>
    /// <param name="semanticModel">The semantic model.</param>
    /// <param name="knownSymbols">Known symbols for type checking.</param>
    /// <param name="eventArguments">The extracted event arguments.</param>
    /// <param name="expectedParameterTypes">The expected parameter types.</param>
    /// <param name="senderCanBeOmitted">Whether the sender argument may be omitted (EventHandler-shaped delegates).</param>
    /// <returns><see langword="true" /> if arguments were successfully extracted; otherwise, <see langword="false" />.</returns>
    internal static bool TryGetEventMethodArgumentsFromLambdaSelector(
        InvocationExpressionSyntax invocation,
        SemanticModel semanticModel,
        KnownSymbols knownSymbols,
        out ArgumentSyntax[] eventArguments,
        out ITypeSymbol[] expectedParameterTypes,
        out bool senderCanBeOmitted)
    {
        return TryGetEventMethodArguments(
            invocation,
            semanticModel,
            out eventArguments,
            out expectedParameterTypes,
            out senderCanBeOmitted,
            static (sm, selector) =>
            {
                bool success = sm.TryGetEventTypeFromLambdaSelector(selector, out ITypeSymbol? eventType);
                return (success, eventType);
            },
            knownSymbols);
    }
```

### 5. Update both analyzer call sites

`src/Analyzers/RaiseEventArgumentsShouldMatchEventSignatureAnalyzer.cs` — replace the body of `Analyze` (currently lines 62-80) with:

```csharp
    private static void Analyze(SyntaxNodeAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        InvocationExpressionSyntax invocation = (InvocationExpressionSyntax)context.Node;

        // Check if this is a Raise method call on a Mock<T>
        if (!IsRaiseMethodCall(context.SemanticModel, invocation, knownSymbols))
        {
            return;
        }

        if (!EventSyntaxExtensions.TryGetEventMethodArgumentsFromLambdaSelector(invocation, context.SemanticModel, knownSymbols, out ArgumentSyntax[] eventArguments, out ITypeSymbol[] expectedParameterTypes, out bool senderCanBeOmitted))
        {
            return;
        }

        string eventName = EventSyntaxExtensions.GetEventNameFromSelector(invocation, context.SemanticModel);

        context.ValidateEventArgumentTypes(eventArguments, expectedParameterTypes, senderCanBeOmitted, invocation, Rule, eventName);
    }
```

`src/Analyzers/RaisesEventArgumentsShouldMatchEventSignatureAnalyzer.cs` — replace the body of `Analyze` (currently lines 65-82) identically (its guard is `context.SemanticModel.IsRaisesInvocation(invocation, knownSymbols)`; only the two changed lines differ):

```csharp
    private static void Analyze(SyntaxNodeAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        InvocationExpressionSyntax invocation = (InvocationExpressionSyntax)context.Node;

        if (!context.SemanticModel.IsRaisesInvocation(invocation, knownSymbols))
        {
            return;
        }

        if (!EventSyntaxExtensions.TryGetEventMethodArgumentsFromLambdaSelector(invocation, context.SemanticModel, knownSymbols, out ArgumentSyntax[] eventArguments, out ITypeSymbol[] expectedParameterTypes, out bool senderCanBeOmitted))
        {
            return;
        }

        string eventName = EventSyntaxExtensions.GetEventNameFromSelector(invocation, context.SemanticModel);

        context.ValidateEventArgumentTypes(eventArguments, expectedParameterTypes, senderCanBeOmitted, invocation, Rule, eventName);
    }
```

These are the only two production callers of the chain (verified by search; `EventSetupHandlerShouldMatchEventTypeAnalyzer` does not use these helpers).

### 6. Update docs

- `docs/rules/Moq1202.md` and `docs/rules/Moq1204.md`: in the "Supported event delegate types" section, add non-generic `EventHandler` and note that for `(object sender, TArgs e)`-shaped delegates both the args-only form (Moq supplies the sender) and the full two-argument form are accepted.
- Do **not** touch `src/Analyzers/AnalyzerReleases.Shipped.md` (no new rule IDs; no severity/ID changes).

### A. Update the direct unit tests in `tests/Moq.Analyzers.Test/Common/EventSyntaxExtensionsTests.cs`

These call the changed helpers directly and must be migrated to the new signatures:

1. `GetEventParameterTypes_SingleTypeArgDelegate_ReturnsSingleType` (line 155-168): the `EventHandler<EventArgs>` row's expectation changes (now returns the full `[object, EventArgs]` with `senderCanBeOmitted == true`). Replace with:

```csharp
    [Fact]
    public void TryGetEventParameterTypes_ActionOfT_ReturnsSingleTypeWithoutOptionalSender()
    {
        (ITypeSymbol eventType, KnownSymbols knownSymbols) = GetEventFieldTypeWithKnownSymbols(
            "using System;\nclass C { event Action<double> MyEvent; }", "MyEvent");

        bool success = EventSyntaxExtensions.TryGetEventParameterTypes(eventType, knownSymbols, out ITypeSymbol[] result, out bool senderCanBeOmitted);

        Assert.True(success);
        Assert.Single(result);
        Assert.Equal("double", result[0].ToDisplayString());
        Assert.False(senderCanBeOmitted);
    }

    [Fact]
    public void TryGetEventParameterTypes_GenericEventHandler_ReturnsInvokeParamsWithOptionalSender()
    {
        (ITypeSymbol eventType, KnownSymbols knownSymbols) = GetEventFieldTypeWithKnownSymbols(
            "using System;\nclass C { event EventHandler<EventArgs> MyEvent; }", "MyEvent");

        bool success = EventSyntaxExtensions.TryGetEventParameterTypes(eventType, knownSymbols, out ITypeSymbol[] result, out bool senderCanBeOmitted);

        Assert.True(success);
        Assert.Equal(2, result.Length);
        Assert.Equal("object", result[0].ToDisplayString());
        Assert.Equal("System.EventArgs", result[1].ToDisplayString());
        Assert.True(senderCanBeOmitted);
    }
```

2. `GetEventParameterTypes_MultiParamDelegate_ReturnsAllParameterTypes` (line 170-185): migrate the call to `TryGetEventParameterTypes(...)`, assert `success` is true and `senderCanBeOmitted` is false; assertions on the two types unchanged.
3. `GetEventParameterTypes_CustomDelegateNoParameters_ReturnsEmpty` (line 187-201): migrate; assert `success` true, result empty, flag false.
4. `GetEventParameterTypes_NonNamedTypeSymbol_ReturnsEmpty` (line 203-223): **behavior change** — rename to `TryGetEventParameterTypes_NonNamedTypeSymbol_ReturnsFalse` and assert the method returns `false` (and outputs an empty array).
5. `GetEventParameterTypes_PlainEventHandler_ReturnsFallbackInvokeParams` (line 225-241): **behavior change** — rename to `TryGetEventParameterTypes_PlainEventHandler_ReturnsInvokeParamsWithOptionalSender`, assert success, 2 parameter types, and `senderCanBeOmitted == true`.
6. New test — unresolved delegate type returns false:

```csharp
    [Fact]
    public void TryGetEventParameterTypes_ErrorType_ReturnsFalse()
    {
        (ITypeSymbol eventType, KnownSymbols knownSymbols) = GetEventFieldTypeWithKnownSymbols(
            "class C { event UnresolvedDelegateType MyEvent; }", "MyEvent");

        Assert.Equal(TypeKind.Error, eventType.TypeKind);

        bool success = EventSyntaxExtensions.TryGetEventParameterTypes(eventType, knownSymbols, out ITypeSymbol[] result, out bool senderCanBeOmitted);

        Assert.False(success);
        Assert.Empty(result);
        Assert.False(senderCanBeOmitted);
    }
```

7. The five `TryGetEventMethodArguments_*` tests (lines 244-406) each gain `out bool _` (insert `out bool senderCanBeOmitted` after `out ITypeSymbol[] expectedParameterTypes` in each call).

### B. New analyzer-level rows in `tests/Moq.Analyzers.Test/RaiseEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs`

Add (file-top `using Microsoft.CodeAnalysis.Testing;` is needed for `CompilerDiagnostics`):

```csharp
    public static IEnumerable<object[]> EventHandlerShapedTestData()
    {
        return new object[][]
        {
            // Canonical Moq pattern for non-generic EventHandler: Moq supplies the sender.
            ["""mock.Raise(n => n.Closed += null, EventArgs.Empty);"""],

            // Two-argument form binds Raise(..., params object[]) and is also legal.
            ["""mock.Raise(n => n.Closed += null, new object(), EventArgs.Empty);"""],

            // EventHandler<CustomEventArgs>: args-only form (existing behavior).
            ["""mock.Raise(n => n.CustomEvent += null, new CustomEventArgs());"""],

            // EventHandler<CustomEventArgs>: two-argument form (previously a false positive).
            ["""mock.Raise(n => n.CustomEvent += null, new object(), new CustomEventArgs());"""],

            // Custom (object sender, TArgs e) delegate: both forms are legal.
            ["""mock.Raise(n => n.ShapedEvent += null, new CustomEventArgs());"""],
            ["""mock.Raise(n => n.ShapedEvent += null, new object(), new CustomEventArgs());"""],

            // Invalid: payload not convertible to EventArgs.
            ["""mock.Raise(n => n.Closed += null, {|Moq1202:42|});"""],

            // Invalid: three arguments for a two-parameter delegate.
            ["""mock.Raise(n => n.Closed += null, new object(), EventArgs.Empty, {|Moq1202:"extra"|});"""],

            // Invalid: no arguments at all.
            ["""{|Moq1202:mock.Raise(n => n.Closed += null)|};"""],
        }.WithNamespaces().WithMoqReferenceAssemblyGroups();
    }

    [Theory]
    [MemberData(nameof(EventHandlerShapedTestData))]
    public async Task ShouldHandleEventHandlerShapedDelegates(string referenceAssemblyGroup, string @namespace, string raiseCall)
    {
        await Verifier.VerifyAnalyzerAsync(
            $$"""
              {{@namespace}}
              using System;

              internal class CustomEventArgs : EventArgs { }

              internal delegate void ShapedEventHandler(object sender, CustomEventArgs e);

              internal interface INotifier
              {
                  event EventHandler Closed;
                  event EventHandler<CustomEventArgs> CustomEvent;
                  event ShapedEventHandler ShapedEvent;
              }

              internal class UnitTest
              {
                  private void Test()
                  {
                      var mock = new Mock<INotifier>();
                      {{raiseCall}}
                  }
              }
              """,
            referenceAssemblyGroup);
    }

    public static IEnumerable<object[]> MultiParameterAndNestedGenericTestData()
    {
        return new object[][]
        {
            // Action<T1, T2> is resolved through the Invoke signature.
            ["""mock.Raise(n => n.PairChanged += null, 1, "two");"""],
            ["""mock.Raise(n => n.PairChanged += null, 1, {|Moq1202:2|});"""],

            // Nested generic payload.
            ["""mock.Raise(n => n.ItemsChanged += null, new List<Dictionary<string, int>>());"""],
            ["""mock.Raise(n => n.ItemsChanged += null, {|Moq1202:new List<int>()|});"""],

            // EventHandler<T> where T does not derive from EventArgs: full form required.
            ["""mock.Raise(n => n.StringHandler += null, new object(), "payload");"""],
            ["""{|Moq1202:mock.Raise(n => n.StringHandler += null, "payload")|};"""],
        }.WithNamespaces().WithMoqReferenceAssemblyGroups();
    }

    [Theory]
    [MemberData(nameof(MultiParameterAndNestedGenericTestData))]
    public async Task ShouldHandleMultiParameterAndNestedGenericDelegates(string referenceAssemblyGroup, string @namespace, string raiseCall)
    {
        await Verifier.VerifyAnalyzerAsync(
            $$"""
              {{@namespace}}
              using System;
              using System.Collections.Generic;

              internal interface INotifier
              {
                  event Action<int, string> PairChanged;
                  event Action<List<Dictionary<string, int>>> ItemsChanged;
                  event EventHandler<string> StringHandler;
              }

              internal class UnitTest
              {
                  private void Test()
                  {
                      var mock = new Mock<INotifier>();
                      {{raiseCall}}
                  }
              }
              """,
            referenceAssemblyGroup);
    }

    public static IEnumerable<object[]> UnresolvableDelegateTestData()
    {
        return new object[][]
        {
            // The delegate type does not resolve (mid-edit code): no Moq1202 must be reported.
            ["""mock.Raise(n => n.Broken += null, 42);"""],
            ["""mock.Raise(n => n.Broken += null, EventArgs.Empty, "extra");"""],
        }.WithNamespaces().WithMoqReferenceAssemblyGroups();
    }

    [Theory]
    [MemberData(nameof(UnresolvableDelegateTestData))]
    public async Task ShouldNotReportWhenEventDelegateTypeDoesNotResolve(string referenceAssemblyGroup, string @namespace, string raiseCall)
    {
        // CompilerDiagnostics.None suppresses CS0246 for the intentionally-unresolved delegate type,
        // mirroring the malformed-code pattern used in MockRepositoryVerifyAnalyzerTests.cs:327.
        await Verifier.VerifyAnalyzerAsync(
            $$"""
              {{@namespace}}
              using System;

              internal interface IBrokenNotifier
              {
                  event UnresolvedDelegateType Broken;
              }

              internal class UnitTest
              {
                  private void Test()
                  {
                      var mock = new Mock<IBrokenNotifier>();
                      {{raiseCall}}
                  }
              }
              """,
            referenceAssemblyGroup,
            CompilerDiagnostics.None);
    }
```

### C. Mirror rows for Moq1204 in `tests/Moq.Analyzers.Test/RaisesEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs`

Add an equivalent data set + test method using the setup-chained pattern; representative rows (template must declare `void Submit();` plus the same three events as in B):

```csharp
    // Valid: canonical args-only form for non-generic EventHandler.
    ["""mockProvider.Setup(x => x.Submit()).Raises(x => x.Closed += null, EventArgs.Empty);"""],

    // Valid: full two-argument form.
    ["""mockProvider.Setup(x => x.Submit()).Raises(x => x.Closed += null, new object(), EventArgs.Empty);"""],

    // Invalid: wrong payload type.
    ["""mockProvider.Setup(x => x.Submit()).Raises(x => x.Closed += null, {|Moq1204:42|});"""],
```

And an unresolved-delegate negative test for Moq1204 identical in shape to B's `ShouldNotReportWhenEventDelegateTypeDoesNotResolve` (with `Setup(x => x.Submit()).Raises(...)` and `CompilerDiagnostics.None`).

### D. Existing tests that must keep passing (unchanged)

- `tests/Moq.Analyzers.Test/RaiseEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs` — all existing rows, including `mock.Raise(n => n.CustomEvent += null, {|Moq1202:"wrong"|})` (same span via the offset path) and the `Action` rows.
- `tests/Moq.Analyzers.Test/RaisesEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs` — all existing rows.
- `tests/Moq.Analyzers.Test/Common/EventSyntaxExtensionsTests.cs` analyzer-pipeline theories (`TooFewArgumentsData` through `RaiseWithEventNameData`) — all unchanged: `TwoParamDelegate(string a, int b)` is not EventHandler-shaped (first parameter is `string`), so its too-few row still flags.
- `tests/Moq.Analyzers.Test/Common/NamedTypeSymbolExtensionsTests.cs` — untouched.

Moq-version note: rows multiply across 2 namespaces and both Moq versions via `WithNamespaces().WithMoqReferenceAssemblyGroups()`. Both Moq 4.8.2 and 4.18.4 expose the same `Raise(Action<T>, EventArgs)` / `Raise(Action<T>, params object[])` and `IRaise<T>.Raises(Action<T>, EventArgs)` / `Raises(Action<T>, params object[])` overloads (verified against both packages' XML docs), so no version-specific splitting is needed.

## Constraints

- Analyzer assemblies target `netstandard2.0`; Roslyn pinned to Microsoft.CodeAnalysis 4.8 (ADR-003); no new package dependencies.
- Symbol-based detection (ADR-001): the new `System.EventHandler`/`System.EventArgs` checks go through `KnownSymbols` (`src/Common/WellKnown/KnownSymbols.cs`) via `TypeProvider.GetOrCreateTypeByMetadataName` — no string comparisons of type names.
- Tests follow repo patterns: `{|Moq1202:...|}` markup asserting both ID and span, `[Theory]` + `[MemberData]`, rows multiplied across 2 namespaces and 2 Moq versions (4.8.2/4.18.4); malformed-code tests use `CompilerDiagnostics.None`.
- Validation: `dotnet format`, `dotnet build` (0 warnings), `dotnet test --settings ./build/targets/tests/test.runsettings`.
- Update `docs/rules/Moq1202.md` and `docs/rules/Moq1204.md` (behavior changes); never touch `AnalyzerReleases.Shipped.md`.

## Validation commands

```bash
dotnet format
dotnet build            # must complete with 0 warnings
dotnet test --settings ./build/targets/tests/test.runsettings
```

## Acceptance criteria

- [ ] `mock.Raise(m => m.Closed += null, EventArgs.Empty)` on `event EventHandler Closed;` produces no Moq1202 (both Moq 4.8.2 and 4.18.4, both namespace variants).
- [ ] The two-argument form on `EventHandler`, `EventHandler<TArgs : EventArgs>`, and custom `(object, TArgs)` delegates produces no Moq1202/Moq1204.
- [ ] Wrong payload types and wrong argument counts on those events are still flagged with the same span conventions as existing tests.
- [ ] An event whose delegate type is an error type produces no Moq1202/Moq1204 regardless of argument count (verified with `CompilerDiagnostics.None`).
- [ ] `EventHandler<T>` with non-`EventArgs` `T` requires the full two-argument form (1-arg form flagged as too few).
- [ ] All pre-existing tests in the four test files listed in "Tests > D" pass without span changes (except the two direct unit tests explicitly listed as behavior changes).
- [ ] `TryGetEventParameterTypes` returns `false` (not an empty array) for non-named types, error types, and non-delegates; unit tests cover all three.
- [ ] `docs/rules/Moq1202.md` and `docs/rules/Moq1204.md` document the EventHandler-shaped acceptance; `AnalyzerReleases.Shipped.md` untouched.
- [ ] `dotnet format`, `dotnet build` (0 warnings), and the full test suite pass.

## Out of scope

- The pre-existing subtle FN where a 1-argument `Raise` call with an `EventArgs`-typed argument on a **non**-shaped delegate (e.g. `event Action<EventArgs>`) binds Moq's `EventArgs` overload and fails at runtime; modeling C# overload resolution of `Raise` itself is beyond this fix.
- `Func<...>`-based `Raises` overloads (`Raises<T1..>(Action<T>, Func<..., EventArgs>)`) — not analyzed today, unchanged here.
- Performance items for these analyzers (per-invocation `GetSymbolInfo` without a cheap pre-filter and eager `GetEventNameFromSelector`) — tracked separately by the audit (A-12/L7).
- Deduplicating the two Raise/Raises analyzer shells (audit E-3).
